File: app/src/main/java/br/com/algorit/manga_alert/retrofit/callback/CallbackWithReturn.java
package br.com.algorit.manga_alert.retrofit.callback;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.internal.EverythingIsNonNull;

public class CallbackWithReturn<T> implements Callback<T> {

    private final ResponseCallback<T> callback;

    public CallbackWithReturn(ResponseCallback<T> callback) {
        this.callback = callback;
    }

    @Override
    @EverythingIsNonNull
    public void onResponse(Call<T> call, Response<T> response) {
        if (response.isSuccessful()) {
            T returned = response.body();
            if (returned != null) {
                callback.success(returned);
            }
        } else {
            callback.fail("Erro ao verificar");
        }
    }

    @Override
    @EverythingIsNonNull
    public void onFailure(Call<T> call, Throwable t) {
        callback.fail("Falha: " + t.getMessage());
    }

    public interface ResponseCallback<T> {
        void success(T response);

        void fail(String fail);
    }
}

File: app/src/main/java/br/com/algorit/mangaalert/retrofit/callback/CallbackWithReturn.kt
package br.com.algorit.mangaalert.retrofit.callback

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.internal.EverythingIsNonNull

class CallbackWithReturn<T>(private val callback: ResponseCallback<T>) : Callback<T> {
    @EverythingIsNonNull
    override fun onResponse(call: Call<T>, response: Response<T>) {
        if (response.isSuccessful) {
            val returned = response.body()
            if (returned != null) {
                callback.success(returned)
            }
        } else {
            callback.fail("Erro ao verificar")
        }
    }

    @EverythingIsNonNull
    override fun onFailure(call: Call<T>, t: Throwable) {
        callback.fail("Falha: " + t.message)
    }

    interface ResponseCallback<T> {
        fun success(response: T)
        fun fail(fail: String)
    }
}